dali/kernels/imgproc/convolution/separable_convolution_gpu.h

```diff
@@ -0,0 +1,179 @@
+ *        If channel axis is pressent, the convolution is not applied there.
+ *        If it is marqed as sequence, the first data axis is considered as temporal axis
+ *
+ * Sequence convolutions require one less window to be specified
+ *
+ * `W` is currently assumed to be float
+ *
+ * `windows` and `scales` are specified per axis.
+ *
+ * Specialized for 1, 2 or 3 axes, to not go overboard with TMP for generic solutions
+ *
+ * Here be boilerplate.
+ */
+template <typename Out, typename In, typename W, int axes, bool has_channels = false,
+          bool is_sequence = false>
+struct SeparableConvolutionGpu;
+
+template <typename Out, typename In, typename W, bool has_channels, bool is_sequence>
+struct SeparableConvolutionGpu<Out, In, W, 1, has_channels, is_sequence> {
+  static constexpr int axes = 1;
+  static constexpr int sequence_axes = static_cast<int>(is_sequence);
+  static constexpr int channel_axes = static_cast<int>(has_channels);
+  static constexpr int ndim = sequence_axes + axes + channel_axes;
+
+  KernelRequirements Setup(KernelContext& ctx, const TensorListShape<ndim>& in_shape,
+                           const std::array<TensorListShape<1>, axes>& window_sizes) {
+    KernelRequirements req;
+    req.output_shapes.push_back(in_shape);
+
+    auto req_conv = conv_.Setup(ctx, in_shape, window_sizes[0]);
+    req.AddInputSet(req_conv, false);
+
+    return req;
+  }
+
+  void Run(KernelContext& ctx, const TensorListView<StorageGPU, Out, ndim> out,
+           const TensorListView<StorageGPU, const In, ndim>& in,
+           const std::array<TensorListView<StorageCPU, const W, 1>, axes>& windows,
+           const std::array<span<const int>, 1> anchors = {},
+           float scale = 1) {
+    conv_.Run(ctx, out, in, windows[0], anchors[0], scale);
+  }
+
+  ConvolutionGpu<Out, In, W, ndim, sequence_axes + 0, has_channels> conv_;
+};
+
+template <typename Out, typename In, typename W, bool has_channels, bool is_sequence>
+struct SeparableConvolutionGpu<Out, In, W, 2, has_channels, is_sequence> {
+  static constexpr int axes = 2;
+  static constexpr int sequence_axes = static_cast<int>(is_sequence);
+  static constexpr int channel_axes = static_cast<int>(has_channels);
+  static constexpr int ndim = sequence_axes + axes + channel_axes;
+  using Intermediate = decltype(std::declval<W>() * std::declval<In>());
+
+  KernelRequirements Setup(KernelContext& ctx, const TensorListShape<ndim>& in_shape,
+                           const std::array<TensorListShape<1>, axes>& window_sizes) {
+    KernelRequirements req;
+
+    ScratchpadEstimator se;
+    se.add<Intermediate>(AllocType::GPU, in_shape.num_elements());
+    req.scratch_sizes = se.sizes;
+    req.output_shapes.push_back(in_shape);
+
+    auto req_inner = conv_innermost_.Setup(ctx, in_shape, window_sizes[1]);
+    auto req_outer = conv_outermost_.Setup(ctx, in_shape, window_sizes[0]);
+
+    req.AddInputSet(req_inner, false);
+    req.AddInputSet(req_outer, false);
```

Comment:
```suggestion
    req.AddInputSet(req_inner, true);
    req.AddInputSet(req_outer, true);
```
Can't you reuse the scratchpad? It should be possible.

Reply:
I see that SeparableConvolutionXXX are the only kernels using this function. And... it's bugeed. Also, it's used incorrectly, since AddInputSet concatenates the list of outputs.

Reply:
Huh, missed that append. I think it was used also in something else, I just wanted to have a way of accumulating the requirements without dealing with the insides of the KernelRequirements representation. Will fix.

Reply:
Done